netbench/netbench-driver/src/bin/netbench-driver-s2n-tls-client.rs

```diff
@@ -0,0 +1,120 @@
+    security::DEFAULT_TLS13,
+};
+use s2n_tls_tokio::{TlsConnector, TlsStream};
+use std::{collections::HashSet, future::Future, net::SocketAddr, pin::Pin, sync::Arc};
+use structopt::StructOpt;
+use tokio::{io::AsyncWriteExt, net::TcpStream};
+
+#[global_allocator]
+static ALLOCATOR: Allocator = Allocator::new();
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    Client::from_args().run().await
+}
+
+#[derive(Debug, StructOpt)]
+pub struct Client {
+    #[structopt(flatten)]
+    opts: netbench_driver::Client,
+}
+
+impl Client {
+    pub async fn run(&self) -> Result<()> {
+        let addresses = self.opts.address_map().await?;
+        let scenario = self.opts.scenario();
+
+        let client = self.client()?;
+        let client = netbench::Client::new(client, &scenario, &addresses);
+        let mut trace = self.opts.trace();
+        let mut checkpoints = HashSet::new();
+        let mut timer = netbench::timer::Tokio::default();
+        client.run(&mut trace, &mut checkpoints, &mut timer).await?;
+
+        Ok(())
+    }
+
+    fn client(&self) -> Result<ClientImpl> {
+        let connector = TlsConnector::new(self.config()?.build()?);
+        let connector: s2n_tls_tokio::TlsConnector = connector.into();
+        let connector = Arc::new(connector);
+
+        let config = multiplex::Config::default();
+
+        Ok(ClientImpl {
+            config,
+            connector,
+            id: 0,
+        })
+    }
+
+    fn config(&self) -> Result<Builder, Error> {
+        let mut builder = Config::builder();
+        builder.set_security_policy(&DEFAULT_TLS13)?;
+        for ca in self.opts.certificate_authorities() {
+            builder.trust_pem(ca.pem.as_bytes())?;
+        }
+        unsafe {
+            builder.disable_x509_verification()?;
+        }
+        Ok(builder)
+    }
+}
+
+type Connection<'a> = netbench::Driver<'a, multiplex::Connection<TlsStream<TcpStream>>>;
+
+#[derive(Clone)]
+struct ClientImpl {
+    config: multiplex::Config,
+    connector: Arc<s2n_tls_tokio::TlsConnector>,
+    id: u64,
+}
+
+impl ClientImpl {
+    fn id(&mut self) -> u64 {
+        let id = self.id;
+        self.id = id + 1;
+        id
+    }
+}
+
+impl<'a> netbench::client::Client<'a> for ClientImpl {
+    type Connect = Pin<Box<dyn Future<Output = Result<Self::Connection>> + 'a>>;
+    type Connection = Connection<'a>;
+
+    fn connect(
+        &mut self,
+        addr: SocketAddr,
+        server_name: &str,
+        server_conn_id: u64,
+        scenario: &'a Arc<scenario::Connection>,
+    ) -> Self::Connect {
+        let id = self.id();
+        let config = self.config.clone();
+        let connector = self.connector.clone();
+        let server_name = server_name.to_string();
+
+        let fut = async move {
+            let conn = TcpStream::connect(addr).await?;
+            let mut conn = connector.connect(&server_name, conn).await?;
+
+            conn.write_u64(server_conn_id).await?;
```

Comment:
We don't need to send the `server_conn_id` with s2n-tls, since it's already included in the `server_name`. Just change the server to read the server name and call `on_server_name` instead.

https://github.com/aws/s2n-quic/blob/43230d34fc7be9823e49105404782008773ba6fa/netbench/netbench-driver/src/bin/netbench-driver-s2n-quic-server.rs#L52-L53

The reason for writing it as part of the stream with native-tls is there's currently no API to retrieve the server name on the server.

Reply:
added in bfcb53b129ccc1d6c3a0e2c92e250c72989a4f3d

It works when testing with https://github.com/aws/s2n-tls/pull/3320.